### ISSUE TITLE
[Explore]: Can we avoid downtime when repushing apps

### ISSUE THREAD

**Issue by danail-branekov** [cloudfoundry/korifi#2479]
### Background

In CF for VMs when a new version of the app is pushed, there is a short downtime period, even if the app has been scaled to more then one instances in advance.

However, in Korifi we could leverage the K8S pod disruption budget, and on a repush, instead of recreating the statefulset, we could just update its image. Then K8S would take care to gradually reschedule the statefulset pods so that no downtime occurs.

### Acceptance Criteria

Proof of concept (and a proposal if controversial) whether updating the statefulset on repush makes sense and does that work.

### Timebox

No timebox

**Comment by kieron-dev** [comment]
## Analysis of current behaviour

From the bottom up...

### Statefulset-runner

#### AppWorkload Controller

This reconciles AppWorkloads with Statefulsets. A Statefulset is created for each AppWorkload. The Statefulset name is constructed from the AppWorkload GUID and Version. The AppWorkload is set to own the Statefulset.

### Korifi Controllers

#### CFProcess Controller

This reconciles CFProcesses with AppWorkloads. It creates AppWorkloads owned by the CFProcesses. The name is constructed from a hash including the CFApp's `app-rev`, stored in an annotation on the CFApp. Thus we have a new AppWorkload each time the CFApp's app-rev changes.

The reconciler also deletes all AppWorkloads which do not match the current app-rev. And if app's desired state is not STARTED or the CFProcess number of desired instances is zero, _all_ AppWorkloads for the process are deleted, including the one for the current app-rev.

#### CFApp Webhook

When a CFApp's `DesiredState` is changed from STARTED to STOPPED, its `app-rev` is incremented.

## What happens during `cf push`

On `cf push`, a new CFPackage is created and source uploaded against it. Then a new CFBuild is created associated with the CFPackage. This triggers a BuildWorkload creation, and eventually a droplet reference bubbles up and is stored on the CFBuild.

Then the CLI sets that droplet reference as the current droplet on the app and restarts the app. This is where the problems begin.

### CFProcess Controller

The restart command first sets the app to STOPPED and then STARTED. These changes might be so quick that a reconciler never sees the STOPPED state, but the CFApp hook will _always_ bump the app-rev.

If the process reconciler triggers while the app is in the STOPPED state, it will delete the AppWorkloads associated with the process. Then when it triggers with the app in the STARTED state, it will create a new AppWorkload with a different name from previously as the app-rev has changed.

If the process reconciler triggered after the transition of the app back to started, it will create a new AppWorkload with a new name (as above), and will then delete the old AppWorkload.

In both cases, the deletion of the AppWorkload causes the running Statefulset to be garbage collected and all instances of the running app go down.

### Statefulset AppWorkload Controller

This will see a new AppWorkload and create a new Statefulset for it.

## The ideal behaviour

The Statefulset controller takes care to set a pod disruption budget on the stateful sets it produces so that when image references change, or pods need to be moved between nodes, a certain number or proportion of pods are left running during the update. If a `cf push` just resulted in a new image reference being set in its spec, the app would update gracefully. Note that changing the image ref is sufficient for the Statefulset to restart all pods, so the extra restart command from the CF CLI is unnecessary.

### Restarting

Since it is unnecessary to restart an app in korifi when its droplet reference is updated, the CF CLI code be modified to omit this step when it knows it is targeting a korifi deployment. Then the app would never enter the STOPPED state, causing AppWorkloads and Statefulsets to be deleted. Also the app-rev would not change, so the AppWorkload with the previous app-rev would not be deleted, and so the Statefulset would persist, and would update gracefully.

We cannot totally ignore restarts in korifi, as they are required for applying updates to env vars, for example. But note that there is a way to gracefully restart pods in a statefulset: `kubectl rollout restart statefulset <statefulset-name>`. If there was a way to tell the statefulset controller that a restart was required, then it could apply this sort of strategy for the restart. A possible way, which goes against the normal declarative kubernetes idiom (but then how does one implement such a command on kubernetes without extending the API?), is to use a property on the AppWorkload to state a restart is required. The controller should perform the restart and clear the field.

**Comment by danail-branekov** [comment]
We discovered that the `kubectl rollout restart` just puts an annotation on the statefulset.spec.template.metadata.annotations map. This causes the statefulset controller to restart the pods gracefully.

So we [spiked](https://github.com/cloudfoundry/korifi/tree/explore-restart-without-downtime) an attempt at implementing graceful restart in korifi. We did following:

1. POST /v3/app/APP_GUID/actions/restart now just sets an annotation of `korifi.cloudfoundry.org/restartedAt: [now]` on the CFApp, and sets DesiredState to STARTED.
2. The CFProcess reconciler copies this annotation to the AppWorkload
3. The CFProcess reconciler now just uses the CFProcess name for the AppWorkload name (it does not use app-rev), so it long-lived.
4. The statefulset-runner's AppWorkload controller now sets the `korifi.cloudfoundry.org/restartedAt` annotation from the AppWorkload onto the statefulset at the spec.template.metadata.annotations level.

This now results in a graceful restart without downtime (when pod count > 1).

We noticed that `cf restart` actual does a stop followed by a start, as does `cf push` and `cf restage`, so we [spiked changing the CLI code too](https://github.com/eirini-forks/cli/tree/explore-restart-without-downtime). This change fixes all 3 commands to restart gracefully!
